tw-tasks-core/src/main/java/com/transferwise/tasks/processing/TasksProcessingService.java

```diff
@@ -700,16 +717,24 @@ public void startProcessing() {
             try {
               long stateVersion = bucket.getVersion().get();
 
-              processTasks(bucket);
+              ProcessTasksResponse processTasksResponse = processTasks(bucket);
+              Instant tryAgainTime = processTasksResponse.getTryAgainTime();
 
               Lock versionLock = bucket.getVersionLock();
               versionLock.lock();
               try {
-                while (bucket.getVersion().get() == stateVersion && !shuttingDown) {
-                  try {
-                    bucket.getVersionCondition().await(tasksProperties.getGenericMediumDelay().toMillis(), TimeUnit.MILLISECONDS);
-                  } catch (InterruptedException e) {
-                    log.error(e.getMessage(), e);
+                if (bucket.getVersion().get() == stateVersion && !shuttingDown) {
+                  long defaultWaitTimeMs = tasksProperties.getGenericMediumDelay().toMillis();
+                  long tryAgainWaitTimeMs =
+                      tryAgainTime == null ? Integer.MAX_VALUE : tryAgainTime.toEpochMilli() - TwContextClockHolder.getClock().millis();
+                  long waitTimeMs = Math.min(defaultWaitTimeMs, tryAgainWaitTimeMs);
+
+                  if (waitTimeMs > 0) {
+                    try {
+                      bucket.getVersionCondition().await(waitTimeMs, TimeUnit.MILLISECONDS);
```

Comment:
So we wait for the `waitTimeMs` + what takes for thread to acquire the lock back

Reply:
Yep, but all places under versionLock are very fast.